### PR TITLE
[RGL] Adjust RGL Unity Plugin to use graph API

### PR DESCRIPTION
**Pull request by prybicki**
Note: the PR is awaiting @msz-rai refactor.

### CODE REVIEW

File: Assets/RGLUnityPlugin/Scripts/LowLevelWrappers/RGLGraph.cs [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+            {
+                nodes.Add(nodeHandle);
+                return;
+            }
+
+            if (childGraphs.Count > 0)
+            {
+                Debug.LogWarning("Added node to already connected graph. Removing child graphs...");
+                DisconnectAllChildGraphs();
+            }
```
Comment (prybicki) [inline Assets/RGLUnityPlugin/Scripts/LowLevelWrappers/RGLGraph.cs:306]: Disconnecting all child graphs will make them return constant output on `rgl_graph_run(...)`, which may be surprising if the warning goes unnoticed. An error is harder to omit usually. At the moment, I don't see any drawbacks to throwing an exception. What do you think?

File: Assets/RGLUnityPlugin/Scripts/LowLevelWrappers/RGLGraph.cs [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+        public void Destroy()
+        {
+            DisconnectAllChildGraphs();
+            DisconnectAllParentGraphs();
+            if (nodes.Count > 0)
+            {
+                RGLNativeAPI.GraphDestroy(nodes.First().Node);
+                nodes.Clear();
+            }
+        }
```
Comment (prybicki) [inline Assets/RGLUnityPlugin/Scripts/LowLevelWrappers/RGLGraph.cs:256]: This `Destroy()` is not destructive, so perhaps it might be named `Clear()`?

**Review by prybicki — COMMENTED** [review]
A number of suggestions are to be discussed. Otherwise, the code looks very clean and safe!

File: Assets/AWSIM/Scripts/Sensors/LiDAR/PointCloudFormats/FormatPCL24.cs [hunk 3]
```diff
@@ -0,0 +1,74 @@
+using System.Collections.Generic;
+using RGLUnityPlugin;
+using sensor_msgs.msg;
+
+namespace AWSIM.PointCloudFormats
```
[the remark this answers is not in the document]
Comment (prybicki) [reply Assets/AWSIM/Scripts/Sensors/LiDAR/PointCloudFormats/FormatPCL24.cs:5]: Note: I'm suggesting moving all formats into a single file below.

File: Assets/RGLUnityPlugin/Scripts/LidarSensor.cs [hunk 4]
```diff
@@ -152,29 +146,41 @@ public void FixedUpdate()
...
+        public void ConnectToWorldFrame(RGLGraph graph)
+        {
+            RGLGraph.ConnectGraphs(rglGraphBase, graph);
         }
 
-        public OutputData RequestCapture()
+        public void ConnectToLidarFrame(RGLGraph graph)
+        {
+            RGLGraph.ConnectGraphs(rglGraphToLidarFrame, graph);
+        }
```
Comment (prybicki) [inline Assets/RGLUnityPlugin/Scripts/LidarSensor.cs:164]: This is beautiful :1st_place_medal:

File: Assets/AWSIM/Scenes/Main/PointCloudMapping/Scripts/Mapper/RGLMappingAdapter.cs [hunk 5]
```diff
@@ -38,32 +38,34 @@ public string GetSensorName()
...
+            // {
+            //     return pcl;
+            // }
+            //
+            // var filteredPCL = new PointCloudOfXYZI();
+            // var voxelGrid = new PclSharp.Filters.VoxelGridOfXYZI();
+            // voxelGrid.SetInputCloud(pcl);
+            // voxelGrid.LeafSize = new PointXYZ { V = new System.Numerics.Vector3(leafSize, leafSize, leafSize) };
+            // voxelGrid.filter(filteredPCL);
+            // return filteredPCL;
```
[the remark this answers is not in the document]
Comment (msz-rai) [reply Assets/AWSIM/Scenes/Main/PointCloudMapping/Scripts/Mapper/RGLMappingAdapter.cs:68]: Yes, this is a placeholder.

File: Assets/AWSIM/Scripts/Sensors/LiDAR/PointCloudFormats/FormatPCL48.cs [hunk 6]
```diff
@@ -0,0 +1,110 @@
+using System.Collections.Generic;
+using RGLUnityPlugin;
+using sensor_msgs.msg;
+using std_msgs.msg;
+
+namespace AWSIM.PointCloudFormats
+{ 
+    public static class FormatPCL48
```
[the remark this answers is not in the document]
Comment (msz-rai) [reply Assets/AWSIM/Scripts/Sensors/LiDAR/PointCloudFormats/FormatPCL48.cs:8]: OK

[the remark this answers is not in the document]
Comment (msz-rai) [reply Assets/AWSIM/Scripts/Sensors/LiDAR/PointCloudFormats/FormatPCL48.cs:8]: OK

File: Assets/RGLUnityPlugin/Scripts/LowLevelWrappers/RGLGraph.cs [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+        private List<RGLGraph> childGraphs = new List<RGLGraph>();
+        private RGLField outputField = RGLField.UNKNOWN;
+
+        //// MULTIGRAPH OPERATIONS ////
+        public static void ConnectGraphs(RGLGraph parentGraph, RGLGraph childGraph)
+        {
+            if (parentGraph.nodes.Count == 0 || childGraph.nodes.Count == 0)
+            {
+                throw new RGLException("Attempted to connect empty graph!");
+            }
```
[the remark this answers is not in the document]
Comment (msz-rai) [reply Assets/RGLUnityPlugin/Scripts/LowLevelWrappers/RGLGraph.cs:39]: Good catch!

File: Assets/RGLUnityPlugin/Scripts/LowLevelWrappers/RGLGraph.cs (hunk 1 above)
Comment (msz-rai) [reply Assets/RGLUnityPlugin/Scripts/LowLevelWrappers/RGLGraph.cs:306]: We had an offline discussion about graph behavior when adding a new node between already connected graphs. We decided to not support these operations at the moment. I agree that exception is better than log message and unpredicted behavior.

File: Assets/AWSIM/Scripts/Sensors/LiDAR/RglLidarPublisher.cs [hunk 8]
```diff
@@ -15,6 +15,8 @@
 
 using UnityEngine;
 using AWSIM;
+using AWSIM.PointCloudFormats;
+using rcl_interfaces.msg;
```
[the remark this answers is not in the document]
Comment (msz-rai) [reply Assets/AWSIM/Scripts/Sensors/LiDAR/RglLidarPublisher.cs]: `AWSIM.PointCloudFormats` is used to get `FormatPCL24` and `FormatPCL48`
`rcl_interfaces.msg` seems to be not used

File: Assets/AWSIM/Scenes/Main/PointCloudMapping/Scripts/Mapper/RGLMappingAdapter.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (prybicki) [reply Assets/AWSIM/Scenes/Main/PointCloudMapping/Scripts/Mapper/RGLMappingAdapter.cs:68]: CC @mackierx111

**Review by msz-rai — APPROVED** [review]

File: Assets/RGLUnityPlugin/Plugins/Windows/x86_64/boost_filesystem-vc143-mt-x64-1_79.dll.meta [hunk 9]
```diff
@@ -0,0 +1,27 @@
+fileFormatVersion: 2
```
[the remark this answers is not in the document]
Comment (prybicki) [reply Assets/RGLUnityPlugin/Plugins/Windows/x86_64/boost_filesystem-vc143-mt-x64-1_79.dll.meta:1]: CC @mackierx111

**Comment by prybicki** [comment]
Superseded by #34

**Pull request closed without merging**